src/amd_sgpio.c

```diff
@@ -330,167 +330,165 @@ static int _send_sgpio_register(const char *em_buffer_path, void *reg,
-
-	log_info("CFG Register: %08x %08x %08x %08x %08x",
-		 reg[0], reg[1], reg[2], reg[3], reg[4]);
-
-	_dump_sgpio_hdr(&cfg_reg->hdr);
-	_dump_sgpio_req(&cfg_reg->req);
-	_dump_sgpio_cfg(&cfg_reg->cfg);
+	log_debug("%s SGPIO AMD Register: %08x\n", type, amd);
+	log_debug(REG_FMT_2, "initiator", get_sgpio_amd_initiator(&amd),
+		  "polarity", get_sgpio_amd_polarity_flip(&amd));
+	log_debug(REG_FMT_2, "bypass enable", get_sgpio_amd_bypass_enable(&amd),
+		  "return to normal", get_sgpio_amd_return_to_normal(&amd));
 }
 
 static int _write_cfg_register(const char *em_buffer_path,
 			       struct cache_entry *cache, int ibpi)
 {
 	struct config_register cfg_reg;
 
-	_init_sgpio_hdr(&cfg_reg.hdr, 0, sizeof(cfg_reg));
-	_init_sgpio_req(&cfg_reg.req, 0x40, 0x82, SGPIO_REQ_REG_TYPE_CFG, 0, 2);
+	cfg_reg.hdr = _init_sgpio_hdr(0, sizeof(cfg_reg));
+	cfg_reg.req = _init_sgpio_req(0x40, 0x82, SGPIO_REQ_REG_TYPE_CFG,
+				      0, 2);
 
 	if (cache->blink_gen_a)
 		cache->blink_gen_b = ibpi_pattern[ibpi];
 	else
 		cache->blink_gen_a = ibpi_pattern[ibpi];
 
-	_init_sgpio_cfg(&cfg_reg.cfg, 1, cache->blink_gen_a, cache->blink_gen_b,
-			2, 1, 0, 0);
+	cfg_reg.cfg = _init_sgpio_cfg(1, cache->blink_gen_a,
+				      cache->blink_gen_b, 2, 1, 0, 0);
+
+	_dump_sgpio_hdr("CFG", cfg_reg.hdr);
+	_dump_sgpio_req("CFG", cfg_reg.req);
+	_dump_sgpio_cfg("CFG", cfg_reg.cfg);
 
-	_dump_cfg_register(&cfg_reg);
 	return _send_sgpio_register(em_buffer_path, &cfg_reg, sizeof(cfg_reg));
 }
 
-static void _dump_sgpio_tx(sgpio_tx_t *tx)
+static void _dump_sgpio_tx(const char *type, sgpio_tx_t tx)
 {
 	int i;
 
-	log_debug("SGPIO TX Register:\n");
+	log_debug("%s SGPIO TX Register: %08x\n", type, tx);
 	for (i = 0; i < 4; i++) {
 		log_debug("\tdrive %d: error %x, locate %x, activity %x\n", i,
-			  get_error_led(&tx->drive[i]),
-			  get_locate_led(&tx->drive[i]),
-			  get_activity_led(&tx->drive[i]));
+			  get_error_led(&tx.drive[i]),
+			  get_locate_led(&tx.drive[i]),
+			  get_activity_led(&tx.drive[i]));
 	}
 }
 
-static void _dump_tx_register(struct transmit_register *tx_reg)
-{
-	uint32_t *reg = (uint32_t *)tx_reg;
-
-	log_info("TX Register:  %08x %08x %08x %08x", reg[0], reg[1],
-		 reg[2], reg[3]);
-
-	_dump_sgpio_hdr(&tx_reg->hdr);
-	_dump_sgpio_req(&tx_reg->req);
-	_dump_sgpio_tx(&tx_reg->tx);
-}
-
 static int _write_tx_register(const char *em_buffer_path,
 			      struct transmit_register *tx_reg)
 {
-	_init_sgpio_hdr(&tx_reg->hdr, 0, sizeof(*tx_reg));
-	_init_sgpio_req(&tx_reg->req, 0x40, 0x82, SGPIO_REQ_REG_TYPE_TX, 0, 1);
+	tx_reg->hdr = _init_sgpio_hdr(0, sizeof(*tx_reg));
+	tx_reg->req = _init_sgpio_req(0x40, 0x82, SGPIO_REQ_REG_TYPE_TX,
+				      0, 1);
+
+	_dump_sgpio_hdr("TX", tx_reg->hdr);
+	_dump_sgpio_req("TX", tx_reg->req);
+	_dump_sgpio_tx("TX", tx_reg->tx);
 
-	_dump_tx_register(tx_reg);
-	return _send_sgpio_register(em_buffer_path, tx_reg, sizeof(*tx_reg));
+	return _send_sgpio_register(em_buffer_path, &tx_reg, sizeof(*tx_reg));
```

Comment:
This line should not be changed. Actually you will pass double pointer here.

Reply:
Good catch.

I have updated the branch with the correction.